documentation/src/main/asciidoc/ch07.asciidoc

```diff
@@ -1,3 +1,324 @@
+   as we want the violation to be reported as if it were directly on the property
+   so we pass a `null` node name to `value()`.
+
+A more interesting example is the case of Guava's `Multimap`: we would like to be
+able to validate both the keys and the values of this container type.
+
+Let's first consider the case of the values. A value extractor extracting them
+is required:
+
+[[example-valueextraction-guava-multimap-value-valueextractor]]
+.A `ValueExtractor` for `Multimap` values
+====
+[source, JAVA, indent=0]
+----
+include::{sourcedir}/org/hibernate/validator/referenceguide/chapter07/valueextractor/MultimapValueValueExtractor.java[tags=include]
+----
+====
+
+It allows to validate constraints for the values of the `Multimap`:
+
+.Constraints on the values of a `Multimap`
+====
+[source, JAVA, indent=0]
+----
+include::{sourcedir}/org/hibernate/validator/referenceguide/chapter07/Container.java[tags=multimapValues]
+----
+====
+
+Another value extractor is required to be able to put constraints on the keys
+of a `Multimap`:
+
+[[example-valueextraction-guava-multimap-key-valueextractor]]
+.A `ValueExtractor` for `Multimap` keys
+====
+[source, JAVA, indent=0]
+----
+include::{sourcedir}/org/hibernate/validator/referenceguide/chapter07/valueextractor/MultimapKeyValueExtractor.java[tags=include]
+----
+====
+
+Once these two value extractors are registered, you can declare constraints on the
+keys and values of a `Multimap`:
+
+.Constraints on the keys and values of a `Multimap`
+====
+[source, JAVA, indent=0]
+----
+include::{sourcedir}/org/hibernate/validator/referenceguide/chapter07/Container.java[tags=multimapKeysAndValues]
+----
+====
+
+The differences between the two value extractors may be a bit subtle at a first
+glance so let's shed some light on them:
+
+ * The `@ExtractedValue` annotation marks the targeted type argument (either
+   `K` or `V` in this case).
+ * We use different node names (`<multimap key>` vs. `<multimap value>`).
+ * In one case, we pass the values to the receiver (third argument of the
+   `keyedValue()` call), in the other, we pass the keys.
+
+Depending on your container type, you should choose the `ValueReceiver`
+method fitting the best:
+
+`value()`::
+    for a simple wrapping container - it is used for ``Optional``s
+`iterableValue()`::
+    for an iterable container - it is used for ``Set``s
+`indexedValue()`::
+    for a container containing indexed values - it is used for ``List``s
+`keyedValue()`::
+    for a container containing keyed values - it is used for ``Map``s
+
+For all these methods, you need to pass a node name: it is the name included in
+the node added to the property path of the constraint violation. As mentioned
+earlier, if the node name is `null`, no node is added to the property path:
+it is be useful for pure wrapper types similar to `Optional`.
+
+The choice of the method used is important as it adds important information to the
```

Comment:
"adds important information" => "adds valuable information"? You already said "important" just a few words before. Alternatively, you could explain why it's important. "adds contextual information", maybe.